cmake/modules/SIMDExt.cmake

```diff
@@ -15,26 +15,27 @@
 #
 
 if(CMAKE_SYSTEM_PROCESSOR MATCHES "aarch64|AARCH64")
-  set(HAVE_ARM 1) 
+  set(HAVE_ARM 1)
   set(save_quiet ${CMAKE_REQUIRED_QUIET})
   set(CMAKE_REQUIRED_QUIET true)
   include(CheckCXXSourceCompiles)
 
   check_cxx_source_compiles("
     #define CRC32CX(crc, value) __asm__(\"crc32cx %w[c], %w[c], %x[v]\":[c]\"+r\"(crc):[v]\"r\"(value))
-    asm(\".arch_extension crc\");
```

Comment:
Why are you removing this line? We need this for gcc 4.8 compatibility, as there gcc does not have any CPU target type that supports crc. But our binutils does support them, so we enable it in the assembly output.

Reply:
@agraf the master branch can only built with GCC-7 and up. so i guess, we don't need to be compatible with lower versions of GCC anymore.

Reply:
In that case, can't we just turn the whole model upside down and just require all features we want to potentially make use of? That would get rid of the error prone "compiler doesn't support feature" code paths and would only leave us with only runtime checks.

Reply:
@agraf by "turning the whole model upside down", do you mean, we should enable all features we are interested, and detect/enable them at runtime?

Reply:
Currently the configure script tries to make the whole world happy. That leaves us with 4 potential different code paths:

  1) feature not available by compiler, but available in CPU
  2) feature not available by compiler, not available in CPU
  3) feature available by compiler, available in CPU
  4) feature available by compiler, not available in CPU

If gcc-7 supports all features we care about, can we just reduce the matrix to only 3 and 4? That way we would always compile the same code everywhere.

We need the runtime check anyway, so nothing changes on that side.

Reply:
So basically configure would fail if -march does not include crypto features for example :). But at least we'd have a completely consistent code base.

The last nasty bug I remember in that field was caused by exactly that discrepancy. There this code:

  https://github.com/ceph/ceph/blob/master/src/common/crc32c.cc#L31

ends up calling ceph_crc32c_aarch64() even in case 1 above. However, if the compiler does not support crc32c extensions, you got at return 0 from the accelerated case.

My point is that if we can make this code as easy as possible by just failing cmake when the right features are not enabled, we'd get much cleaner test coverage.

Reply:
My opinion is that the compiled code should always match march/mcpu, but if you want to use runtime feature detection instead (completely ignoring march/mcpu) then there are bugs with the current implementation that lead to https://tracker.ceph.com/issues/23464

A side-note: GCC-7 (GCC 6 and up AFAIK) implements ARM ACLE 2.0 (https://gcc.gnu.org/onlinedocs/gcc-7.3.0/gcc/ARM-C-Language-Extensions-_0028ACLE_0029.html) sufficiently well that all ARM features have #defines provided by the compiler, for example `__ARM_FEATURE_CRYPTO`, `__ARM_FEATURE_CRC32` etc (see http://infocenter.arm.com/help/topic/com.arm.doc.ihi0053c/IHI0053C_acle_2_0.pdf). If compile-time checks are still desired in any way, this might be a better option than the current CMake stuff.

Reply:
Compile time feature detection is really only useful in unsupportable Gentoo-like environments. Once you get to packages shipped in binary distributions, features have to be determined during runtime.

The real question behind this is: Do you want to figure out you missed out on 20% performance because you didn't enable the right compiler flags?

I think most people don't. They would much rather prefer the configure script / cmake to bail out and say "you're not compiling with an optimal configuration, please give me a good CPU model to work against".

If you really want to have something that allows you to disable CPU features statically, I'd much rather see them as opt-out rather than dynamically determined.

Reply:
Also please bear in mind that I'm not 100% sure -march is the right way to decide these things either, as that would allow gcc to emit native instructions for a machine it may not run on.

What we really need is something that indicates to the compiler "I know what I'm doing, let me drive the intrinsics because I have a runtime check behind them".

Reply:
> The real question behind this is: Do you want to figure out you missed out on 20% performance because you didn't enable the right compiler flags?

Yes. This is how my other software works, and it's how I expect Ceph to work. The compiler should do what I tell it to do, and if that includes disabling features that are normally available on my machine (because the target machine doesn't have those features) then that's exactly what I expect to happen.

I still think that compiler flags should never be ignored, especially the target march/mcpu settings. I agree with your point about binary distributions though, and if you can get reliable runtime detection of features then that's clearly a better approach. But it's clear that's broken at the moment, since I encountered the problem in the linked error report.

See https://github.com/ceph/ceph/blob/0603794d533da644593184c4a77ef442e021f6cc/src/common/crc32c_aarch64.c#L98: this function is selected if the CRC feature is available on the CPU, but the feature checks do not check for the crypto feature which is also needed here. Thus a CPU with CRC but not crypto will crash will the illegal instruction error.